### PR TITLE
Support multi trial jobs on same GPU

### PR DESCRIPTION
**Pull request by SparkSnail**

### CODE REVIEW

File: src/nni_manager/rest_server/restValidationSchemas.ts [hunk 1]
```diff
@@ -31,10 +31,14 @@ export namespace ValidationSchemas {
                 passwd: joi.string(),
                 sshKeyPath: joi.string(),
                 passphrase: joi.string(),
-                gpuIndices: joi.string()
+                gpuIndices: joi.string(),
+                maxTrialNumPerGPU: joi.number(),
+                useActiveGPU: joi.boolean()
```
Comment (chicm-ms) [inline src/nni_manager/rest_server/restValidationSchemas.ts]: For gpuNum and gpuIndices, 'gpu' is lower case,  suggest to use 'maxTrialNumPerGpu', 'useActiveGpu' for consistency.
And better to add document for these 2 new configuration like gpuNum and gpuIndices.

Comment (QuanluZhang) [reply src/nni_manager/rest_server/restValidationSchemas.ts]: agree

File: src/nni_manager/training_service/local/localTrainingService.ts [hunk 2]
```diff
@@ -117,13 +125,15 @@ class LocalTrainingService implements TrainingService {
...
-    private occupiedGpuIndices: Set<number>;
+    private occupiedGpuIndexNumMap: Map<number, number>;
     private designatedGpuIndices!: Set<number>;
     private log: Logger;
     private localTrailConfig?: TrialConfig;
     private localConfig?: LocalConfig;
     private isMultiPhase: boolean = false;
     private jobStreamMap: Map<string, ts.Stream>;
+    private maxTrialNumPerGPU: number = 1;
+    private useActiveGPU: boolean = false;
```
Comment (QuanluZhang) [inline src/nni_manager/training_service/local/localTrainingService.ts]: how about init them in constructor?

File: src/nni_manager/training_service/remote_machine/gpuScheduler.ts [hunk 3]
```diff
@@ -87,21 +88,32 @@ export class GPUScheduler {
...
-        if (rmMeta !== undefined && rmMeta.gpuReservation !== undefined) {
-            rmMeta.gpuReservation.forEach((reserveTrialJobId : string, gpuIndex : number) => {
-                if (reserveTrialJobId === trialJobId) {
-                    rmMeta.gpuReservation.delete(gpuIndex);
+    public removeGpuReservation(trialJobId: string, trialJobMap: Map<string, RemoteMachineTrialJobDetail>): void {
+        let trialJobDetail: RemoteMachineTrialJobDetail | undefined = trialJobMap.get(trialJobId);
+        if(trialJobDetail === undefined) {
+            throw new Error(`could not get trialJobDetail by id ${trialJobId}`);
+        } 
+        if (trialJobDetail.rmMeta !== undefined && trialJobDetail.rmMeta.occupiedGpuIndexMap !== undefined && trialJobDetail.gpuIndices !== undefined && trialJobDetail.gpuIndices.length > 0) {
```
Comment (QuanluZhang) [inline src/nni_manager/training_service/remote_machine/gpuScheduler.ts]: better to split this line to four lines

File: src/nni_manager/training_service/remote_machine/gpuScheduler.ts [hunk 4]
```diff
@@ -145,10 +154,18 @@ export class GPUScheduler {
...
+                    // or trial number on a GPU reach max number,
                     // We should NOT allocate this GPU
-                    if (gpuInfo.activeProcessNum === 0 && !rmMeta.gpuReservation.has(gpuInfo.index)
-                        && (designatedGpuIndices === undefined || designatedGpuIndices.has(gpuInfo.index))) {
-                        availableGPUs.push(gpuInfo);
+                    if (designatedGpuIndices === undefined || designatedGpuIndices.has(gpuInfo.index)) {
+                        if(rmMeta.occupiedGpuIndexMap !== undefined) {
+                            let num = rmMeta.occupiedGpuIndexMap.get(gpuInfo.index);
+                            let maxTrialNumPerGPU: number = rmMeta.maxTrialNumPerGPU? rmMeta.maxTrialNumPerGPU: 1;
+                            if((num === undefined && (!rmMeta.useActiveGPU && gpuInfo.activeProcessNum === 0 || !rmMeta.useActiveGPU)) || (num !== undefined && num < maxTrialNumPerGPU)) {
```
Comment (QuanluZhang) [inline src/nni_manager/training_service/remote_machine/gpuScheduler.ts]: the same, better to use multiple lines

Comment (QuanluZhang) [reply src/nni_manager/training_service/remote_machine/gpuScheduler.ts]: suggest 2 lines for this case

File: src/nni_manager/training_service/remote_machine/gpuScheduler.ts [hunk 5]
```diff
@@ -170,14 +187,23 @@ export class GPUScheduler {
...
         allocatedGPUs.forEach((gpuInfo: GPUInfo) => {
-            rmMeta.gpuReservation.set(gpuInfo.index, trialJobId);
+            if(rmMeta.occupiedGpuIndexMap !== undefined) {
+                let num = rmMeta.occupiedGpuIndexMap.get(gpuInfo.index);
+                if(num === undefined) {
+                    num = 0;
+                }
+                rmMeta.occupiedGpuIndexMap.set(gpuInfo.index, num + 1);
+            }else {
+                rmMeta.occupiedGpuIndexMap = new Map<number, number>();
```
Comment (QuanluZhang) [inline src/nni_manager/training_service/remote_machine/gpuScheduler.ts]: also reconsider this logic

File: src/nni_manager/rest_server/restValidationSchemas.ts (hunk 1 above)
Comment (SparkSnail) [reply src/nni_manager/rest_server/restValidationSchemas.ts]: set to `maxTrialNumPerGpu` and `useActiveGpu`

File: src/nni_manager/training_service/local/localTrainingService.ts (hunk 2 above)
Comment (SparkSnail) [reply src/nni_manager/training_service/local/localTrainingService.ts]: fixed, moved to constructor.

File: src/nni_manager/training_service/remote_machine/gpuScheduler.ts (hunk 3 above)
Comment (SparkSnail) [reply src/nni_manager/training_service/remote_machine/gpuScheduler.ts]: fixed, split to four lines.

File: src/nni_manager/training_service/remote_machine/gpuScheduler.ts (hunk 4 above)
Comment (SparkSnail) [reply src/nni_manager/training_service/remote_machine/gpuScheduler.ts]: fixed, split to 2 lines.

File: src/nni_manager/training_service/remote_machine/gpuScheduler.ts (hunk 5 above)
Comment (SparkSnail) [reply src/nni_manager/training_service/remote_machine/gpuScheduler.ts]: fixed, initialize in constructor.

File: tools/nni_cmd/launcher.py [hunk 6]
```diff
@@ -160,9 +160,13 @@ def set_local_config(experiment_config, port, config_file_name):
     request_data = dict()
     if experiment_config.get('localConfig'):
         request_data['local_config'] = experiment_config['localConfig']
-        if request_data['local_config'] and request_data['local_config'].get('gpuIndices') \
-            and isinstance(request_data['local_config'].get('gpuIndices'), int):
-            request_data['local_config']['gpuIndices'] = str(request_data['local_config'].get('gpuIndices'))
+        if request_data['local_config']:
+            if request_data['local_config'].get('gpuIndices') and isinstance(request_data['local_config'].get('gpuIndices'), int):
+                request_data['local_config']['gpuIndices'] = str(request_data['local_config'].get('gpuIndices'))
+            if request_data['local_config'].get('maxTrialNumOnEachGPU'):
```
Comment (QuanluZhang) [inline tools/nni_cmd/launcher.py]: GPU?

File: tools/nni_cmd/launcher.py [hunk 7]
```diff
@@ -160,9 +160,13 @@ def set_local_config(experiment_config, port, config_file_name):
...
+                request_data['local_config']['maxTrialNumOnEachGPU'] = request_data['local_config'].get('maxTrialNumOnEachGPU')
+            if request_data['local_config'].get('useActiveGPU'):
+                request_data['local_config']['useActiveGPU'] = request_data['local_config'].get('useActiveGPU')
```
Comment (QuanluZhang) [inline tools/nni_cmd/launcher.py]: GPU? -> Gpu

File: tools/nni_cmd/launcher.py (hunk 6 above)
Comment (SparkSnail) [reply tools/nni_cmd/launcher.py]: fixed.

File: tools/nni_cmd/launcher.py (hunk 7 above)
Comment (SparkSnail) [reply tools/nni_cmd/launcher.py]: fixed.

File: docs/en_US/ExperimentConfig.md [hunk 8]
```diff
@@ -399,6 +399,15 @@ machineList:
   
     __gpuIndices__ is used to specify designated GPU devices for NNI, if it is set, only the specified GPU devices are used for NNI trial jobs. Single or multiple GPU indices can be specified, multiple GPU indices are seperated by comma(,), such as `1` or  `0,1,3`.
 
+  * __maxTrialNumPerGpu__
+  
+    __maxTrialNumPerGpu__ is used to specify the max concurrency trial number on a GPU device.
+    
+  * __useActiveGpu__
+  
+    __useActiveGpu__ is used to specify whether to use a GPU if there is another process. By default, NNI will use the GPU only if there is no another active process in the GPU, if __useActiveGpu__ is set to true, NNI will use the GPU regardless of another processes.
```
Comment (chicm-ms) [inline docs/en_US/ExperimentConfig.md]: need to state that this option is not applicable for NNI on Windows.

Comment (SparkSnail) [reply docs/en_US/ExperimentConfig.md]: Fixed.

File: docs/en_US/ExperimentConfig.md [hunk 9]
```diff
@@ -433,6 +442,14 @@ machineList:
...
     __gpuIndices__ is used to specify designated GPU devices for NNI on this remote machine, if it is set, only the specified GPU devices are used for NNI trial jobs. Single or multiple GPU indices can be specified, multiple GPU indices are seperated by comma(,), such as `1` or  `0,1,3`.
 
+  * __maxTrialNumPerGpu__
+  
+    __maxTrialNumPerGpu__ is used to specify the max concurrency trial number on a GPU device.
+
+  * __useActiveGpu__
+  
+    __useActiveGpu__ is used to specify whether to use a GPU if there is another process. By default, NNI will use the GPU only if there is no another active process in the GPU, if __useActiveGpu__ is set to true, NNI will use the GPU regardless of another processes. This field is not applicable for NNI on Windows.
+
```
Comment (leelaylay) [inline docs/en_US/ExperimentConfig.md]: Nice Job!
I test it on my local machine and it works successfully.
It seems that it also works in remote mode when I catch a sight of the code.

I suggest put one example code here.

```yaml
localConfig:
  maxTrialNumPerGpu:  5
```

Comment (SparkSnail) [reply docs/en_US/ExperimentConfig.md]: good suggestion, I've added an configuration in cifar10-pytorch-example

**Pull request merged**